src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php

```diff
@@ -39,98 +38,81 @@ public function process(ContainerBuilder $container)
         $this->compiler = $container->getCompiler();
         $this->formatter = $this->compiler->getLoggingFormatter();
 
-        foreach ($container->getAliases() as $id => $alias) {
+        $seenAliases = array();
+        $replacements = array();
+        foreach ($container->getAliases() as $defId => $alias) {
             $aliasId = (string) $alias;
-
+            if (isset($replacements[$aliasId])) {
```

Comment:
the easiest way to handle it would probably be to restart the handling of aliases to build the table of replacements, keeping the update of Definitions for the end to do it only once.